components/tikv_alloc/src/jemalloc.rs

```diff
@@ -21,18 +27,125 @@ lazy_static! {
         Mutex::new(HashMap::new());
 }
 
+thread_local! {
+    static MEMORY_STAT_ACCESSOR_STILL_ALIVE: AccessorStillAlive = {
+        let thd = std::thread::current();
+        let name = thd.name().unwrap_or("<unknown>").to_owned();
+        // We must store the current thread ID here,
+        // or once we are running the destructor,
+        // we probably cannot access these anymore.
+        AccessorStillAlive(name, thd.id())
+    };
+}
+
+struct AccessorStillAlive(String, ThreadId);
+
+impl Drop for AccessorStillAlive {
+    fn drop(&mut self) {
+        let mut l = match THREAD_MEMORY_MAP.lock() {
+            Ok(l) => l,
+            Err(_) => {
+                // perhaps we are panicking, don't panic again (which leads to abort).
+                eprintln!("<tikv_alloc>: the global accessor has been poisoned!");
+                return;
+            }
+        };
+
+        if let Some(accessor) = l.get_mut(&self.1) {
+            // NOTE: should we notify the code writer here? For example adding a
+            // `debug_panic` here?
+            eprintln!(
+                "<tikv_alloc>: the thread {}({:?}) exits with accessor left. forgot to call `remove_thread_memory_accessor`?",
+                self.0, self.1,
+            );
+            accessor.allocated.0 = None;
+            accessor.deallocated.0 = None;
+        }
+    }
+}
+
+/// The struct for tracing the statistic of another thread.
+/// The target pointer should be bound to some TLS of another thread, this
+/// structure is just "peeking" it -- with out modifying.
+// It should be covariant so we wrap it with `NonNull`.
+#[repr(transparent)]
+struct PeekableRemoteStat<T>(Option<NonNull<T>>);
+
+// SAFETY: we need to keep the thread called add_thread_memory_accessor always
```

Comment:
What will happen if it doesn't? I see `MemoryStatsAccessor` is not a thread local but a normal heap object.

Reply:
The pointee of this is a thread local variable(If I correctly understand the `tsd` implementation in jemalloc.). So it become dangling while the remote thread exits.
For now, the safety is guarded by the auto-removal-when-dropping thread local variable `MEMORY_STAT_ACCESSOR_STILL_ALIVE`. I guess that is somehow isomorphic to storing `MemoryStatsAccessor` in TLS. I will check that.

Reply:
(the current implementation of `THREAD_MEMORY_MAP` is a sort of home-made thread local storage...)

Reply:
`ThreadLocal` that I mentioned is also a userspace TLS implementation. It supports traversal so you don't need to implement your own map anymore.

> The pointee of this is a thread local variable(If I correctly understand the tsd implementation in jemalloc.).

Needs to be more specific in comments, e.g. `allocatedp::read()` returns a thread-local.